primitives/src/market.rs

```diff
@@ -69,16 +69,121 @@ pub struct Market<AI, BA, BN, M, A> {
+                } else {
+                    false
+                }
+            }
+            OutcomeReport::Scalar(_) => {
+                matches!(&self.market_type, MarketType::Scalar(_))
+            }
+        }
+    }
+
+    pub fn outcome_assets<MI: Copy + HasCompact + MaxEncodedLen>(
+        &self,
+        market_id: MI,
+    ) -> Vec<MarketAssetClass<MI>> {
+        match self.market_type {
+            MarketType::Categorical(categories) => {
+                let mut assets = Vec::new();
+
+                for i in 0..categories {
+                    match self.scoring_rule {
+                        ScoringRule::Orderbook => {
+                            assets.push(MarketAssetClass::<MI>::CategoricalOutcome(market_id, i))
+                        }
+                        ScoringRule::Lmsr => {
+                            assets.push(MarketAssetClass::<MI>::CategoricalOutcome(market_id, i))
+                        }
+                        ScoringRule::Parimutuel => {
+                            assets.push(MarketAssetClass::<MI>::ParimutuelShare(market_id, i))
+                        }
+                    };
+                }
+
+                assets
+            }
+            MarketType::Scalar(_) => {
+                vec![
+                    MarketAssetClass::<MI>::ScalarOutcome(market_id, ScalarPosition::Long),
+                    MarketAssetClass::<MI>::ScalarOutcome(market_id, ScalarPosition::Short),
+                ]
+            }
+        }
+    }
+
+    pub fn report_into_asset<MI: HasCompact + MaxEncodedLen>(
+        &self,
+        market_id: MI,
+    ) -> Option<MarketAssetClass<MI>> {
+        let outcome = if let Some(ref report) = self.report {
+            &report.outcome
+        } else {
+            return None;
+        };
+
+        self.outcome_report_into_asset(market_id, outcome)
+    }
+
+    pub fn resolved_outcome_into_asset<MI: HasCompact + MaxEncodedLen>(
+        &self,
+        market_id: MI,
+    ) -> Option<MarketAssetClass<MI>> {
+        let outcome = self.resolved_outcome.as_ref()?;
+        self.outcome_report_into_asset(market_id, outcome)
+    }
+
+    fn outcome_report_into_asset<MI: HasCompact + MaxEncodedLen>(
+        &self,
+        market_id: MI,
+        outcome_report: &OutcomeReport,
+    ) -> Option<MarketAssetClass<MI>> {
+        match outcome_report {
+            OutcomeReport::Categorical(idx) => match self.scoring_rule {
+                ScoringRule::Orderbook => {
+                    Some(MarketAssetClass::<MI>::CategoricalOutcome(market_id, *idx))
+                }
+                ScoringRule::Lmsr => {
+                    Some(MarketAssetClass::<MI>::CategoricalOutcome(market_id, *idx))
+                }
+                ScoringRule::Parimutuel => {
+                    Some(MarketAssetClass::<MI>::ParimutuelShare(market_id, *idx))
+                }
+            },
+            OutcomeReport::Scalar(_) => None,
```

Comment:
Actually, one more thing. I'm kind of confused by this one. Why do scalar markets return `None`? `report_into_asset` is not documented, so I'm not sure if this is expected behavior, but it sure is confusing.

Reply:
Added docstrings for the new public functions: https://github.com/zeitgeistpm/zeitgeist/pull/1281/commits/de8944a535baba0bd60e89018d5b0afa5f56c7c9

The reason it returns `None` is that no asset exists that represents the `OutcomeReport`. To be more precise, the `OutcomeReport::Scalar(u128)` provides a precise number of the outcome, whereas the underlying assets used for trading are `MarketAsset::ScalarOutcome(id, Long/Short)`.